### PR TITLE
Factor out sample loading logic from Thunder and Specimen into a separate file

### PR DESCRIPTION
**Pull request by PoroCYon**
Makes porting stuff etc. easier later on.

### CODE REVIEW

File: WaveSabreCore/src/SampleLoader.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	HACMDRIVERID SampleLoader::driverId = NULL;
+
+	SampleLoader::LoadedSample SampleLoader::LoadSampleGSM(char *data, int compressedSize, int uncompressedSize, WAVEFORMATEX *waveFormat)
+	{
+		LoadedSample ret;
+
+		ret.compressedSize = compressedSize;
+		ret.uncompressedSize = uncompressedSize;
+
+		//if (waveFormatData) delete [] waveFormatData;
```
Comment (kusma) [inline WaveSabreCore/src/SampleLoader.cpp:16]: Seeing all these commented out delete-statements, makes me think this code is unfinished...

By the way, @alkama has some variations on this code that uses libsndfile on macOS. I hope we aren't making things difficult for him. I also have [a simplified but untested version](https://github.com/kusma/WaveSabre/tree/macos-libsndfile-cleanups) of that code that I was meaning to upstream, that threads in the same space.

Comment (PoroCYon) [reply WaveSabreCore/src/SampleLoader.cpp:16]: Those were notes I should've kept only temporarily when I was moving the code from Thunder/Specimen and reworking their LoadSample functions. I'll remove it.

And yeah, some coordination with alkama would be great. the only difficulty might be in making the commits work with one another, but there's nothing technically or conceptually difficult, afaict.

Comment (kusma) [reply WaveSabreCore/src/SampleLoader.cpp]: No, I'm not worried here or anything. We can figure out the divergence once we land something

**Comment by yupferris** [comment]
Quick note on why I haven't accepted this yet: it looks good, but I need to make sure it doesn't increase compressed size (by an unreasonable amount). A (competent) user will never use `Specimen` and `Thunder` in the same song (they shouldn't use `Thunder` at all!), so this decoupling will likely only negatively affect code size from a user's perspective.

**Comment by PoroCYon** [comment]
I suppose I could change things a bit so that the `Specimen.LoadSample`/`Thunder.LoadSample` functions *aren't* inline, and then make `SampleLoader::LoadSampleGSM` inline so that with a competent compiler, the binary output will be more or less the same as before these changes.

Does that sound good?

**Comment by kusma** [comment]
[`__forceinline`](https://docs.microsoft.com/en-us/cpp/cpp/inline-functions-cpp?redirectedfrom=MSDN&view=vs-2019#inline-__inline-and-__forceinline) might be what you want for that...

**Comment by PoroCYon** [comment]
Here's a commit that does that. If it's desireable, I'll squash it into the other one.

**Comment by yupferris** [comment]
I finally got around to testing this size-wize (after rebase); using Reverence.als from the converter tests and squishy 1.3.0, current master is 29056 bytes, whereas this branch is 29068 bytes. As expected, slightly worse, but only by 12 bytes, which is quite OK imo! I have some other nits that I think I'll test out next, and we'll see where we end up then, but so far, looking good.

**Comment by yupferris** [comment]
So for an additional 12 bytes (lol), I was able to simplify things further by refactoring `SampleLoader` into a `GsmSample` class instead, which unifies a lot of the memory management code (now the devices just manage a single `sample` instance, and the various other buffers etc are handled internally to that). You can see my changes on [this branch](https://github.com/yupferris/WaveSabre/commits/sample-loading-refactor). Thoughts?

**Comment by yupferris** [comment]
Superceded by #72.

**Pull request closed without merging**